File: 346. Moving Average from Data Stream/346. Moving Average from Data Stream.js
/**
 * Initialize your data structure here.
 * @param {number} size
 */
var MovingAverage = function(size) {
  this.size = size;
  this.q = [];
  this.avg = 0;
};

/**
 * @param {number} val
 * @return {number}
 */
MovingAverage.prototype.next = function(val) {
  if (this.size > this.q.length) {
    this.q.push(val);
    let sum = 0;
    for (let num of this.q) sum += num;
    this.avg = sum / this.q.length;
  } else {
    let head = this.q.shift();
    let minus = head / this.size;
    this.q.push(val);
    let add = val / this.size;
    this.avg = this.avg - minus + add;
  }
  return this.avg;
};

/**
 * Your MovingAverage object will be instantiated and called as such:
 * var obj = new MovingAverage(size)
 * var param_1 = obj.next(val)
 */

File: 346. Moving Average from Data Stream/346. Moving Average from Data Stream.java
import java.util.LinkedList;
import java.util.Queue;
class MovingAverage {
    Queue<Integer> q;
    int size;
    double avg;
    /** Initialize your data structure here. */
    public MovingAverage(int size) {
        this.size = size;
        this.q = new LinkedList<>();
    }
    
    public double next(int val) {
        if(this.size > q.size()){
            q.offer(val);
            int sum = 0;
            for(int num : q){
                sum += num;
            }
            avg = (double) sum / q.size();
        }else{
            int head = q.poll();
            double minus = (double) head /this.size;
            q.offer(val);
            double add = (double) val / this.size;
            avg = avg - minus + add;
        }
        return avg;
    }
}

/**
 * Your MovingAverage object will be instantiated and called as such:
 * MovingAverage obj = new MovingAverage(size);
 * double param_1 = obj.next(val);
 */